opt_einsum/path_random.py

```diff
@@ -0,0 +1,320 @@
+import math
+import time
+import heapq
+import random
+import functools
+from collections import deque
+
+from . import helpers
+from .paths import PathOptimizer, _ssa_optimize, _BETTER_FNS, ssa_to_linear, _calc_k12_flops
+
+
+__all__ = ["RandomGreedy", "random_greedy"]
+
+
+class RandomOptimizer(PathOptimizer):
+    """Base class for running any random path finder that benefits
+    from repeated calling, possibly in a parallel fashion. Custom random
+    optimizers should subclass this, and the ``setup`` method should be
+    implemented with the following signature::
+
+        def setup(self, inputs, output, size_dict):
+            # custom preparation here ...
+            return trial_fn, trial_args
+
+    Where ``trial_fn`` itself should have the signature::
+
+        def trial_fn(r, *trial_args):
+            # custom computation of path here
+            return ssa_path, cost, size
+
+    Where ``r`` is the run number and could for example be used to seed a
+    random number generator. See ``RandomGreedy`` for an example.
+
+
+    Parameters
+    ----------
+    max_repeats : int, optional
+        The maximum number of repeat trials to have.
+    max_time : float, optional
+        The maximum amount of time to run the algorithm for.
+    minimize : {'flops', 'size'}, optional
+        Whether to favour paths that minimize the total estimated flop-count or
+        the size of the largest intermediate created.
+    executor : executor-pool like, optional
+        An executor-pool to optionally parallelize repeat trials over. The pool
+        should have an api matching those found in the python 3 standard library
+        module ``concurrent.futures``. Namely, a ``submit`` method that returns
+        ``Future`` objects, themselves with ``result`` and ``cancel`` methods.
+    pre_dispatch : int, optional
+        If using an ``executor``, how many jobs to pre-dispatch so as to avoid
+        submitting all jobs at once. Should also be more than twice the number
+        of workers to avoid under-subscription. Default: 128.
+
+    Attributes
+    ----------
+    path : list[tuple[int]]
```

Comment:
Reminds me that we should look into a Py2 deprecation plan and think about Py3+ (and typing!).